.azure-pipelines/ultimate-pipeline.yml

```diff
@@ -898,6 +945,144 @@ stages:
     - script: tracer\build.cmd CheckBuildLogsForErrors
       displayName: Check logs for errors
 
+- stage: exploration_tests_linux
+  condition: and(succeeded(), eq(variables['isScheduledBuild'], 'False'), or(eq(stageDependencies.check_changes.outputs['check_changes_job.check_changes_step.IsProfilerChanged'],'True'), eq(stageDependencies.check_changes.outputs['check_changes_job.check_changes_step.IsDebuggerChanged'], 'True')))
+  dependsOn: [build_linux, check_changes]
+  jobs:
+  - template: steps/update-github-status-jobs.yml
+    parameters:
+      jobs: [Test]
+
+  - job: generator
+    variables:
+      ISDEBUGGERCHANGED: $[stageDependencies.check_changes.check_changes_job.outputs['check_changes_step.IsDebuggerChanged']]
+      ISPROFILERCHANGED: $[stageDependencies.check_changes.check_changes_job.outputs['check_changes_step.IsProfilerChanged']]
+    pool:
+      vmImage: ubuntu-18.04
+    steps:
+      - bash: |
+          matrix="{"
+
+          declare -a explorationTestUseCases=()
+          if [ $ISDEBUGGERCHANGED == "True" ]; then 
+            explorationTestUseCases+=("debugger")
+          fi
+
+          if [ $ISPROFILERCHANGED == "True" ]; then 
+            explorationTestUseCases+=("continuousProfiler")
+          fi
+
+          add_to_matrix () {
+            name="${baseImage//./_}_${publishTargetFramework//./_}_${explorationTestUseCase//./_}_${explorationTestName}"
+            matrix="${matrix} '${name}':"
+            matrix="${matrix} { 'baseImage': '${baseImage}', "
+            matrix="${matrix} 'publishTargetFramework': '${publishTargetFramework}', "
+            matrix="${matrix} 'explorationTestUseCase': '${explorationTestUseCase}', "
+            matrix="${matrix} 'explorationTestName': '${explorationTestName}' },"
+          }
+
+          for baseImage in "debian" "alpine" ; do
+            for explorationTestUseCase in ${explorationTestUseCases[@]} ; do
+
+              publishTargetFramework="netcoreapp2.1"
+              for explorationTestName in "eshoponweb" "protobuf" ; do
+                add_to_matrix
+              done
+
+              publishTargetFramework="netcoreapp3.1"
+              explorationTestName="cake"
+              add_to_matrix
+
+              publishTargetFramework="net5.0"
+              explorationTestName="cake"
+              add_to_matrix
+
+              publishTargetFramework="net6.0"
+              for explorationTestName in  "cake" "swashbuckle"; do
```

Comment:
Bit of a shame that we now have more than one place to indicate where each exploration tests can run, would have been nicer if it could have ALL inside the `ExplorationTestDescription`. I don't know NUKE well, so this may be completely irrelevant, but is there no way to use [NUKE Configuration Generation](https://nuke.build/docs/authoring-builds/ci-integration.html) to generate these tasks?

Reply:
I don't like the situation, when we are required to to set those conditions twice, too.

First of -it doesn't solve the problem, since `NUKE Configuration Generation` generates files, so we'll still will be required to `check in` files.

Second, it will be complicated to adopt, since we are using one pipeline for all tasks, so configurating attribute across all other partial classes, sounds tempting.

The solution I can think about - storing `TestDescription` not hardly coded, but in json/yml config file, so both, code and pipeline will use same source, and only one file will be required for change.

Reply:
Yeah, I'm not a fan of this amount of logic going into bash either. The matrix generators were meant to be super-basic loops, as soon as we start having logic like this in them, I think they should be defined in Nuke somewhere. 

One non-optimal solution might be to generate all the permutations in this generator, and then just skip the ones that don't apply in Nuke. Pro: simple bash, compatibility defined in one location (Nuke). Con: lots of "wasted" times/hosts for runs that won't continue.

A preferable approach IMO is to generate this list in Nuke. It can then be called from both locations, is defined in one place, doesn't use bash at all, and only runs permutations that are valid. Downside is that the generator step becomes slightly slower, but I think that's likely still preferable.

Reply:
We can move **all** parameters generations to nuke, and make them run in parallel with the "build" steps.
So, running time will be no longer an issue.

Reply:
> We can move **all** parameters generations to nuke, and make them run in parallel with the "build" steps. So, running time will be no longer an issue.

Oooh, nice idea, I like it!

Reply:
I've moved generations of matrix strategies to nuke build (except of smoke tests).
Please, review.